diff --git a/meson.build b/meson.build
index 14d0955..82b90b8 100644
--- a/meson.build
+++ b/meson.build
@@ -67,10 +67,6 @@ defines = [
   '-D_DEFAULT_SOURCE',
 ]
 
-if host_machine.system() == 'darwin'
-  defines += '-DGTKOSXAPPLICATION'
-endif
-
 # compile flags
 flags = [
   '-Werror=implicit-function-declaration',
@@ -153,6 +149,11 @@ headers = files(
 )
 headers += version_header
 
+if host_machine.system() == 'darwin'
+  sources += files('zathura/mac-integration.c')
+  headers += files('zathura/mac-integration.h')
+endif
+
 # zathura helper library
 libzathura = static_library('zathura',
   sources,
diff --git a/zathura/mac-integration.c b/zathura/mac-integration.c
new file mode 100644
index 0000000..e1c867f
--- /dev/null
+++ b/zathura/mac-integration.c
@@ -0,0 +1,73 @@
+/* SPDX-License-Identifier: Zlib */
+
+#include <girara/log.h>
+#include <glib.h>
+#include <gtkosxapplication.h>
+
+#include "mac-integration.h"
+
+#include "plugin.h"
+#include "shortcuts.h"
+#include "zathura.h"
+
+
+static void load_plugins_from_bundle(zathura_t* zathura);
+static gboolean openfile_callback(GtkosxApplication* app, gchar* path, zathura_t* zathura);
+
+
+void zathura_mac_init(zathura_t* zathura) {
+  load_plugins_from_bundle(zathura);
+
+  GtkosxApplication* zathura_mac = g_object_new(GTKOSX_TYPE_APPLICATION, NULL);
+  gtkosx_application_set_use_quartz_accelerators(zathura_mac, FALSE);
+  gtkosx_application_ready(zathura_mac);
+
+#if 0
+  gchar* id = gtkosx_application_get_bundle_id();
+  if (id != NULL) {
+    girara_warning("TestIntegration Error! Bundle has ID %s", id);
+  }
+  g_free(id);
+#endif
+
+  g_signal_connect(G_OBJECT(zathura_mac), "NSApplicationOpenFile", G_CALLBACK(openfile_callback), zathura);
+}
+
+static void load_plugins_from_bundle(zathura_t* zathura) {
+  zathura_plugin_manager_t* plugin_manager = zathura->plugins.manager;
+
+  gchar* resource_path = gtkosx_application_get_resource_path();
+  gchar* plugins_path = g_build_path("/", resource_path, "plugins", NULL);
+
+  zathura_plugin_manager_set_dir(plugin_manager, plugins_path);
+  zathura_plugin_manager_load(plugin_manager);
+
+  g_free(resource_path);
+  g_free(plugins_path);
+}
+
+static gboolean openfile_callback(GtkosxApplication* UNUSED(app), gchar* path, zathura_t* zathura) {
+  if (zathura_has_document(zathura) == false) {
+    document_open_idle(zathura, path, NULL, ZATHURA_PAGE_NUMBER_UNSPECIFIED, NULL, NULL, NULL, NULL);
+    return true;
+  }
+
+  pid_t pid = fork();
+
+  if (pid == 0) { /* child */
+    if (setsid() == -1) {
+      girara_error("Could not start new process group: %s", strerror(errno));
+    }
+
+    char *args[] = { gtkosx_application_get_executable_path(), path, NULL };
+    execv(args[0], args);
+
+    girara_error("Could not replaces process image: %s", strerror(errno));
+    exit(1);
+  } else if (pid < 0) { /* error */
+    girara_error("Could not fork: %s", strerror(errno));
+    /* Should we somehow notify user about this situation? */
+  }
+
+  return true;
+}
diff --git a/zathura/mac-integration.h b/zathura/mac-integration.h
new file mode 100644
index 0000000..45da763
--- /dev/null
+++ b/zathura/mac-integration.h
@@ -0,0 +1,16 @@
+/* SPDX-License-Identifier: Zlib */
+
+#ifndef MAC_INTEGRATION_H
+#define MAC_INTEGRATION_H
+
+#include "zathura.h"
+
+/**
+ * Initializes zathura OSX integration
+ *
+ * @param zathura The zathura session
+ * @return zathura OSX session object or NULL if zathura could not be created
+ */
+void zathura_mac_init(zathura_t* zathura);
+
+#endif /* MAC_INTEGRATION_H */
diff --git a/zathura/main.c b/zathura/main.c
index f326b4b..fab4c03 100644
--- a/zathura/main.c
+++ b/zathura/main.c
@@ -1,9 +1,5 @@
 /* SPDX-License-Identifier: Zlib */
 
-#ifdef GTKOSXAPPLICATION
-#include <gtkosxapplication.h>
-#endif
-
 #include <girara/settings.h>
 #include <girara/log.h>
 
@@ -18,6 +14,9 @@
 #include "zathura.h"
 #include "plugin.h"
 #include "utils.h"
+#ifdef MAC_INTEGRATION
+#include "mac-integration.h"
+#endif
 #ifdef WITH_SYNCTEX
 #include "dbus-interface.h"
 #include "synctex.h"
@@ -282,17 +281,10 @@ GIRARA_VISIBLE int main(int argc, char* argv[]) {
     goto free_and_ret;
   }
 
-#ifdef GTKOSXAPPLICATION
-  GtkosxApplication* zathuraApp = g_object_new(GTKOSX_TYPE_APPLICATION, NULL);
-  gtkosx_application_set_use_quartz_accelerators(zathuraApp, FALSE);
-  gtkosx_application_ready(zathuraApp);
-  {
-    const gchar* id = gtkosx_application_get_bundle_id();
-    if (id != NULL) {
-      girara_warning("TestIntegration Error! Bundle has ID %s", id);
-    }
-  }
-#endif // GTKOSXAPPLICATION
+#ifdef MAC_INTEGRATION
+  zathura_mac_init(zathura);
+#endif
+
   /* run zathura */
   gtk_main();
 
